src/Microsoft.AspNetCore.SystemWebAdapters/Authentication/RemoteAppAuthenticationModule.Framework.cs

```diff
@@ -2,44 +2,96 @@
+            throw new ArgumentOutOfRangeException(nameof(options.RemoteServiceOptions.ApiKey), "Options must specify API key.");
+        }
+
+        if (string.IsNullOrEmpty(options.RemoteServiceOptions.ApiKeyHeader))
+        {
+            throw new ArgumentOutOfRangeException(nameof(options.RemoteServiceOptions.ApiKeyHeader), "Options must specify API key header name.");
+        }
+
         _options = options;
+        _remoteAppAuthHandler = new RemoteAppAuthenticationHttpHandler();
     }
 
     public void Init(HttpApplication context)
     {
-        var handler = new RemoteAppAuthenticationHttpHandler();
-
-        context.PostMapRequestHandler += MapRemoteAuthenticationHandler;
-
-        void MapRemoteAuthenticationHandler(object sender, EventArgs e)
+        context.PostMapRequestHandler += (s, _) =>
         {
-            var context = ((HttpApplication)sender).Context;
-
+            var context = ((HttpApplication)s).Context;
             if (string.Equals(context.Request.Path, _options.AuthenticationEndpointPath, StringComparison.OrdinalIgnoreCase)
                 && context.Request.HttpMethod.Equals("GET", StringComparison.OrdinalIgnoreCase))
             {
-                if (!string.Equals(_options.RemoteServiceOptions.ApiKey, context.Request.Headers.Get(_options.RemoteServiceOptions.ApiKeyHeader), StringComparison.Ordinal))
+                MapRemoteAuthenticationHandler(new HttpContextWrapper(context));
+
+                if (context.Handler is null)
                 {
-                    // Using 407 here (proxy authentication required) to differentiate from the scenario of
-                    // a valid API key but no authenticated user.
-                    context.Response.StatusCode = 407;
                     context.ApplicationInstance.CompleteRequest();
                 }
-                else
-                {
-                    context.Handler = handler;
-                }
             }
+        };
+    }
+
+    public void MapRemoteAuthenticationHandler(HttpContextBase context)
+    {
+        var apiKey = context.Request.Headers.Get(_options.RemoteServiceOptions.ApiKeyHeader);
+        if (apiKey is null)
+        {
+            // A missing API key may indicate that instead of being a request from the ASP.NET Core app,
+            // this request represents a redirect back from an external authentication provider (for example,
+            // in OIDC or WS-Fed scenarios). In those scenarios, the external authentication provider may attempt
+            // to redirect back to the URL that initiated the login request.
+            //
+            // In those scenarios, if there is an original-url query parameter present in the request's URL, we
+            // can redirect back to that as the *actual* original URL that the authentication process needs to
+            // redirect back to.
+            var originalUrl = context.Request.QueryString[AuthenticationConstants.OriginalUrlQueryParamName];
```

Comment:
do we need to do any sort of data protection on this? i.e. could this be used in a nefarious way?

Reply:
Hmm that makes sense. It wouldn't end up bypassing anything or disclosing any information, right? It's probably fine then

Reply:
Right